contracts/src/v0.8/llo-feeds/FeeManager.sol

```diff
@@ -0,0 +1,353 @@
+  /// @param newSurcharge Surcharge amount to apply relative to PERCENTAGE_SCALAR
+  event NativeSurchargeUpdated(uint256 newSurcharge);
+
+  /// @notice Emits when this contract does not have enough LINK to send to the reward manager when paying in native
+  /// @param configDigest Config digest of the report
+  /// @param linkQuantity Amount of LINK required to pay the reward
+  /// @param nativeQuantity Amount of native required to pay the reward
+  event InsufficientLink(bytes32 indexed configDigest, uint256 linkQuantity, uint256 nativeQuantity);
+
+  /// @notice Emitted when funds are withdrawn
+  /// @param assetAddress Address of the asset withdrawn
+  /// @param quantity Amount of the asset withdrawn
+  event Withdraw(address adminAddress, address assetAddress, uint256 quantity);
+
+  /**
+   * @notice Construct the FeeManager contract
+   * @param _linkAddress The address of the LINK token
+   * @param _nativeAddress The address of the wrapped ERC-20 version of the native token (represents fee in native or wrapped)
+   * @param _proxyAddress The address of the proxy contract
+   * @param _rewardManagerAddress The address of the reward manager contract
+   */
+  constructor(
+    address _linkAddress,
+    address _nativeAddress,
+    address _proxyAddress,
+    address _rewardManagerAddress
+  ) ConfirmedOwner(msg.sender) {
+    if (
+      _linkAddress == address(0) ||
+      _nativeAddress == address(0) ||
+      _proxyAddress == address(0) ||
+      _rewardManagerAddress == address(0)
+    ) revert InvalidAddress();
+
+    i_linkAddress = _linkAddress;
+    i_nativeAddress = _nativeAddress;
+    i_proxyAddress = _proxyAddress;
+    i_rewardManager = IRewardManager(_rewardManagerAddress);
+
+    IERC20(i_linkAddress).approve(address(i_rewardManager), type(uint256).max);
+  }
+
+  modifier onlyOwnerOrProxy() {
+    require(msg.sender == owner() || msg.sender == i_proxyAddress, "Only owner or proxy");
+    _;
+  }
+
+  /// @inheritdoc TypeAndVersionInterface
+  function typeAndVersion() external pure override returns (string memory) {
+    return "FeeManager 0.0.1";
+  }
+
+  /// @inheritdoc IERC165
+  function supportsInterface(bytes4 interfaceId) external pure override returns (bool) {
+    return interfaceId == this.processFee.selector;
+  }
+
+  /// @inheritdoc IFeeManager
+  function processFee(bytes calldata payload, address subscriber) external payable onlyOwnerOrProxy {
+    if (subscriber == address(this)) revert InvalidAddress();
+
+    //decode the report from the payload
+    (, bytes memory report) = abi.decode(payload, (bytes32[3], bytes));
+
+    //get the feedId from the report
+    bytes32 feedId = bytes32(report);
+
+    //v1 doesn't need a quote payload, so skip the decoding if the report is a v0 report
+    Quote memory quote;
+    if (getReportVersion(feedId) != REPORT_V1) {
+      //all reports greater than v0 should have a quote payload
+      (, , , , , bytes memory quoteBytes) = abi.decode(
```

Comment:
Is this smart enough to not decode data it doesn't need?

Reply:
It won't deserialize the whole payload, it will figure out which parts to skip based on the types passed into the parameter